File: src/tools/query.rs
use oxigraph::sparql::QueryResults;
use rmcp::model::CallToolResult;
use serde_json::{json, Value};

use crate::store::{MemoryStore, StoreError};

pub fn handle(store: &MemoryStore, sparql: String) -> Result<CallToolResult, StoreError> {
    let results = store.query(&sparql)?;

    let output = match results {
        QueryResults::Solutions(solutions) => {
            let variables: Vec<String> = solutions
                .variables()
                .iter()
                .map(|v| v.as_str().to_string())
                .collect();

            let mut rows: Vec<Value> = Vec::new();
            for solution in solutions {
                let solution = solution.map_err(|e| StoreError::Other(e.to_string()))?;
                let mut row = serde_json::Map::new();
                for var in &variables {
                    if let Some(term) = solution.get(var.as_str()) {
                        row.insert(var.clone(), json!(term.to_string()));
                    }
                }
                rows.push(Value::Object(row));
            }
            serde_json::to_string_pretty(&rows).unwrap_or_else(|_| "[]".to_string())
        }
        QueryResults::Boolean(b) => format!("{b}"),
        QueryResults::Graph(triples) => {
            let mut lines = Vec::new();
            for triple in triples {
                let triple = triple.map_err(|e| StoreError::Other(e.to_string()))?;
                lines.push(format!(
                    "{} {} {} .",
                    triple.subject, triple.predicate, triple.object
                ));
            }
            lines.join("\n")
        }
    };

    Ok(CallToolResult::success(vec![rmcp::model::Content::text(
        output,
    )]))
}

File: claude_memory_graph/tools/query.py
import json

import pyoxigraph as ox

from ..store import MemoryStore
from ..namespaces import MEM, RDF, RDFS, XSD


_PREFIXES = [(MEM, "mem:"), (RDF, "rdf:"), (RDFS, "rdfs:"), (XSD, "xsd:")]


def _term(term) -> str:
    if isinstance(term, ox.NamedNode):
        for iri, prefix in _PREFIXES:
            if term.value.startswith(iri):
                return prefix + term.value[len(iri):]
        return term.value
    if isinstance(term, ox.Literal):
        return term.value
    return str(term)


def handle(store: MemoryStore, sparql: str) -> str:
    results = store.query(sparql)

    if isinstance(results, ox.QuerySolutions):
        variables = list(results.variables)
        rows = []
        for solution in results:
            row = {}
            for var in variables:
                term = solution[var]
                if term is not None:
                    row[var.value] = _term(term)
            rows.append(row)
        if not rows:
            return "No results."
        return json.dumps(rows, separators=(",", ":"), ensure_ascii=False)

    if isinstance(results, bool):
        return str(results).lower()

    # CONSTRUCT / DESCRIBE → QueryTriples
    lines = []
    for triple in results:
        lines.append(f"{_term(triple.subject)} {_term(triple.predicate)} {_term(triple.object)} .")
    return "\n".join(lines) if lines else "No results."
